### PR TITLE
feat: travelex

### PR DESCRIPTION
**Pull request by charlesdassonville**

### CODE REVIEW

File: locations/spiders/travelex.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                    "lon": float(row.get("lng")),
+                    "website": row.get("storeUrl"),
+                    "extras": {
+                        "directions": row.get("directions"),
+                        "notes": row.get("notes"),
+                        "terminal": row.get("terminal")
+                    }
+                }
+
+                yield Feature(**properties)
```
Comment (davidhicks) [inline locations/spiders/travelex.py:42]: Use DictParser here? Sample:


from locations.dict_parser import DictParser

item = DictParser.parse(row)
item["whatever"] = row["something"] # Adjust fields that DictParser misses or gets wrong
...
yield item

**Review by davidhicks — CHANGES_REQUESTED** [review]

File: locations/spiders/travelex.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import scrapy
+
+from locations.items import Feature
+
+
+class TravelexSpider(scrapy.Spider):
+    name = "travelex"
```
[the remark this answers is not in the document]
Comment (charlesdassonville) [reply locations/spiders/travelex.py:7]: countries added in travelex.py
and NL added in gwk_travelex.py where spider is renamed travelex_gb

[the remark this answers is not in the document]
Comment (davidhicks) [reply locations/spiders/travelex.py:7]: Great to see the API supports other countries too. Good find :)

File: locations/spiders/gwk_travelex.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 
 class TravelexSpider(scrapy.Spider):
-    name = "travelex"
+    name = "travelex_nl"
```
Comment (davidhicks) [inline locations/spiders/gwk_travelex.py:7]: The spider name should probably be `gwk_travelex_nl` ? The file name and class name should also match.

File: locations/spiders/travelex.py [hunk 4]
```diff
@@ -9,7 +9,7 @@ class TravelexSpider(scrapy.Spider):
     allowed_domains = ["https://www.travelex.co.uk/"]
 
     def start_requests(self):
-        countries = ['dech', 'au', 'gb', 'enbh', 'de', "nl", "zhhk", "jajp", "my", "nz", "qa"]
+        countries = ['dech', 'au', 'gb', 'enbh', 'de', "zhhk", "jajp", "my", "nz", "qa"]
```
Comment (davidhicks) [inline locations/spiders/travelex.py]: `om`, `ae` can be added. There are a few more countries for which I couldn't figure out the right language/country code.

`nl` should probably be added back here again as it uses the same API? I'm not sure what the purpose of splitting gwk_travelex into a separate spider? The `brand` and `brand_wikidata` attributes can then be modified for `nl` as required to use the alternate brand `GWK Travelex`.

Comment (charlesdassonville) [reply locations/spiders/travelex.py]: i'll try that !

File: locations/dict_parser.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 
 class DictParser:
-    ref_keys = ["ref", "id", "store-id", "shop-number", "slug"]
+    ref_keys = ["ref", "id", "store-id", "shop-number", "slug", "storeId"]
```
Comment (Cj-Malone) [inline locations/dict_parser.py:5]: We do variations on these so `storeId` is already included with `store-id`

File: locations/dict_parser.py [hunk 6]
```diff
@@ -86,7 +86,7 @@ class DictParser:
         "yext-display-lng",
     ]
 
-    website_keys = ["url", "website", "permalink"]
+    website_keys = ["url", "website", "permalink", "storeUrl"]
```
Comment (Cj-Malone) [inline locations/dict_parser.py:89]: Yeah, we should handle `storeUrl`, but I'd prefer `store-url` to match the others.

```suggestion
    website_keys = ["url", "website", "permalink", "store-url"]
```

File: locations/spiders/gwk_travelex.py [hunk 7]
```diff
@@ -17,26 +18,22 @@ def start_requests(self):
 
     def parse(self, response):
         data = response.json()
-
         item_categories = data.get("items")
         for category in item_categories:
             stores = category.get("stores")
             for row in stores:
+                item = DictParser.parse(row)
                 properties = {
```
Comment (davidhicks) [inline locations/spiders/gwk_travelex.py]: `properties` can be removed. Just do e.g. `item["postcode"] = row.get("address").get("postalCode"),` instead for simplicity?

File: locations/spiders/travelex.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        data = response.json()
+        item_categories = data.get("items")
+        for category in item_categories:
+            stores = category.get("stores")
+            for row in stores:
+                item = DictParser.parse(row)
+                item["addr_full"] = row.get("formattedAddress")
+                item["street_address"] = row.get("address").get("address1")
+                item["city"] = row.get("address").get("city")
+                item["postcode"] = row.get("address").get("postalCode")
```
Comment (Cj-Malone) [inline locations/spiders/travelex.py:30]: Are these already handled by DictParser? If so:
```suggestion
```

**Comment by Cj-Malone** [comment]
CI gives us [468](https://data.alltheplaces.xyz/map.html?show=https://data.alltheplaces.xyz/ci/4356802215/travelex/output.geojson).

We're getting some bad `addr_full`, eg `Belfast International, Airport Rd\nBelfast\nBT29 4AB`. A `.replace("\n", ", ")` may be enough, or we may want to use `clean_address` if there are any white space issues as well.

**Review by Cj-Malone — APPROVED** [review]
[LGTM](https://data.alltheplaces.xyz/map.html?show=https://data.alltheplaces.xyz/ci/4362609102/travelex/output.geojson), thanks

**Review by davidhicks — CHANGES_REQUESTED** [review]

File: locations/spiders/travelex.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+from locations.spiders.vapestore_gb import clean_address
+
+
+class TravelexSpider(scrapy.Spider):
+    name = "travelex"
+    item_attributes = {"brand": "Travelex", "brand_wikidata": "Q2337964"}
+    allowed_domains = ["https://www.travelex.co.uk/"]
+
+    def start_requests(self):
+        countries = ["dech", "au", "gb", "enbh", "de", "zhhk", "jajp", "my", "nz", "qa", "om", "ae", "nl"]
```
[the remark this answers is not in the document]
Comment (charlesdassonville) [reply locations/spiders/travelex.py]: done

**Pull request merged**